### ISSUE TITLE
Enable specifying classpath to include other submodules for Type Resolution

### ISSUE THREAD

**Issue by Munzey** [detekt/detekt#7808]
## Expected Behavior

I want to be able to specify that other modules in my gradle build should be included as part of classpath but *not*  for analysis, only to assist in type resolution.

It seems the extension only allows me to set [source](https://github.com/detekt/detekt/blob/main/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt#L23) packages

## Current Behavior

If i resolve a type in a class in my module that detekt is applied to:
```kt
bindingContext[BindingContext.TYPE, typeReference]
```
and the type is from other included module `X` then the type class descriptor will be `deserialized class MyOtherModuleClass`

## Context

I want to be able to access PsiElements of types from other modules in my project for further analysis.
Perhaps this is possible now, feel free to move this to a question if so.

**Comment by Munzey** [comment]
For more context - Im running from an Android project and ive been calling `detektDebug` to run.
looking at [the plugin code](https://github.com/detekt/detekt/blob/v1.23.7/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektAndroid.kt#L143) (as of writing from latest release of 1.23.7 - the code has been refactored on `main` since https://github.com/detekt/detekt/pull/7016) - it doesnt seem there is actually a way to set `extraInputSource` for an android task via the detekt extension [source](https://github.com/detekt/detekt/blob/v1.23.7/detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt#L46) param. 

Reading the doc it seems I might need to write my own task based on https://github.com/detekt/detekt/issues/2259

**Comment by Munzey** [comment]
ok best I could come up with was to follow https://github.com/detekt/detekt/issues/2259#issuecomment-669888008
In case this helps someone else:
we have detekt applied per module of a multi module project.  If type resolution is required to resolve the sibling module types then you can add the additional source required like so:

```kt
fun findKotlinSourceFiles(projectName: String): List<File> {
  val projectDir = project(":$projectName").projectDir
  val sourceDir = projectDir.resolve("src/main/java")
  return if (sourceDir.exists()) {
    sourceDir.walkTopDown().filter { it.isFile && it.extension == "kt" }.toList()
  } else {
    emptyList()
  }
}

tasks.withType<Detekt>().configureEach {
  var additionalSourceSet = emptyList<File>()
  additionalSourceSet += findKotlinSourceFiles(projects.yourModuleName.name)
  additionalSourceSet += findKotlinSourceFiles(projects.yourOtherModuleName.name)

  doFirst {
    source(additionalSourceSet)
  }
}
```

however this does have the setback that if you have baselined issues for those other modules - now you will also need to baseline for this module.

Ideally there would be a way to include these other modules on the classpath with their source resolved so that they are not included as part of analysis.

**Comment by cortinico** [comment]
> I want to be able to specify that other modules in my gradle build should be included as part of classpath but _not_ for analysis, only to assist in type resolution.

As you found out, this is not easily possible.

The caveat of using the `SourceTask` is a good workaround, and you're probably fine with that. Also why is it a problem if the extra sources are also used for analysis. Is it for performance reasons?

**Comment by BraisGabin** [comment]
I'm probably missing something here. Detekt should add to the classpath all the modules that are used to compile your module. If those other modules are used they should be added to the classpath too. Also, if two modules are siblings then they don't depend between each other. So it should be fine if they are not added to the classpath. What am I missing?

**Comment by Munzey** [comment]
hey sorry just getting back to this now!

@cortinico 
>Also why is it a problem if the extra sources are also used for analysis. Is it for performance reasons?

it's that if there is any other detekt issues in that additional source code that might already be baselined in their respective modules, then they also get picked up and raised in this module. 
So you might have additional duplicated baselined issues that you need to remember to rerun/remove if they ever get fixed in the other modules.

**Comment by Munzey** [comment]
@BraisGabin 
>Detekt should add to the classpath all the modules that are used to compile your module. If those other modules are used they should be added to the classpath too.

yes they are there and the code compiles and the detekt rules indeed run against the source code of the module using the code from whatever else is on the classpath. However, with type resolution there is a limitation to what can get resolved about the classes you might be depending on from other modules that you own the source of.

I think easiest to explain with my current use case:

I have an Android app comprising of multiple gradle modules that all get included as part of the main build. Each module has detekt plugin applied to it.

I have a use case where I need to ensure that an annotation exists on certain data classes that can exist in any of these modules - specifically any that are referenced as parameters or return types of our retrofit instances.

so ive written a custom rule that checks the module to see if it contains one of these retrofit interfaces, then scans all the function args and return types of this interface to see if the types are annotated.
I do this by resolving the type, then recursively checking all type arguments, constructor args, public properties as well as for nested classes to see if any else also require the annotation.

this works really well as long as all the types are classes in the modules source.
If a referenced data class comes from a different module, then resolving the `KotlinType` results in a descriptor of `deserialized class MyOtherModuleType` that yes with it I can still access everything I need to verify if it has the annotation or not, but I can't resolve it to a `PsiElement` using `DescriptorToSourceUtils.descriptorToDeclaration` due to it being treated as a library/external type. So I can still report the issue, but i have to reference the psielement from the source code module retorfit interface where I started from, which is a lot less clear to the developer who encounters the issue.

There isnt really a way around this as some "core" modules won't have any retrofit instances so the only way to discover if they have data classes that are referenced somewhere else and need the annotation is to go through this module-to-module type resolving process.

hence why I was wondering is there a way we could specify that we want the source included in the compiled artifact from other modules declared as dependencies instead of being treated like any other 3rd party compiled artifact/dependency.

Hope that makes sense! will try clarify anything else if needed.
